RubberduckTests/Grammar/VBAParserTests.cs

```diff
@@ -2953,6 +2953,50 @@ End Sub
             var parseResult = Parse(code);
         }
 
+        [Category("Parser")]
+        [Test]
+        public void ParserDoesNotFailDoubleBracketForgeignIdentifierWithTypeHint()
+        {
+            const string code = @"
+Sub Test()  
+Dim x
+x = [[bar]]!
+End Sub
+";
+            var parseResult = Parse(code);
+            AssertTree(parseResult.Item1, parseResult.Item2, "//typeHint", matches => matches.Count == 1);
+        }
+
+        [Category("Parser")]
+        [Test]
+        public void ParserDoesNotFailOnBangOperatorFollowedByForeignIdentifier()
+        {
+            const string code = @"
+Sub Test()   
+Dim dict As Scripting.Dictionary
+
+Dim x
+x = dict![a]
+End Sub
+";
+            var parseResult = Parse(code);
```

Comment:
Should we also test that the `a` is correctly parsed as an identifier? That may be more important for other cases where we have a expression like `dic![[Test]![Run]]`; the `[Test]![Run]` in this case would be an identifier, not a piece of code.

Reply:
I think that would be a different test and actually be of a different nature than the other tests. None of those test specific parts of the code, just that it parses and that some elements appear the correct number of times. Here, we could test the correct number of `unrestrictedIdentifier`.

BTW, we ne er parse `a` as an identifier: the `unrestrictedIdentifier` in question is `[a]`.

Reply:
That is fine by me - I just want to be sure we were covering the supported edge cases correctly. One more thought - do we need a separate test to assert the count of `unrestrictedIdentifer` for a chained bangs expression like `Forms![Main]![SomeControl]`/`Forms!Main!SomeControl` ?

Reply:
I will add a test for chained bang notation without brackets and an ignored one with brackets. The version with brackets on the left cannot work the way we do this right now.